### PR TITLE
runtime: add Roc filetype

### LINKED ISSUE #14415: Add Roc language filetype
[Roc](https://www.roc-lang.org/) is a fast, friendly, functional language.
Roc uses the `.roc` extension for its files. I am attempting to help make
the experience of developing Roc better by adding support for it in
various tools, and I think having Roc be automatically detected in Vim
would be useful.

### PR DESCRIPTION
**Pull request by nat-418**
Closes https://github.com/vim/vim/issues/14415

### CODE REVIEW

File: runtime/ftplugin/roc.vim [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+" Maintainer: nat-418 <93013864+nat-418@users.noreply.github.com>
+" Latest Revision: 2024-04-5
+
+if exists("b:did_ftplugin")
+  finish
+endif
+let b:did_ftplugin = 1
+
+let s:cpo_save = &cpo
+set cpo&vim
```
Comment (chrisbra) [inline runtime/ftplugin/roc.vim]: I think there is nothing in your plugin requiring the 
```
let s:cpo_save = &cpo
set cpo&vim
```
dance and you can remove it.

File: runtime/ftplugin/roc.vim [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+setlocal comments=:#
+setlocal commentstring=#%s
```
Comment (dkearns) [inline runtime/ftplugin/roc.vim]: I don't know the language's formatting norms but it's probably more common to specify `commenstring` as `#\ %s` for a line comment. 

This is what the Elm ftplugin does, for example.

https://github.com/ElmCast/elm-vim/blob/4b71facd77297cb33bbb3b14894676cff0a9bd1d/ftplugin/elm.vim#L41

Comment (clason) [reply runtime/ftplugin/roc.vim]: Sorry for hijacking this discussion, but it struck me as a good (and relevant) opportunity: Is there an official Vim policy on commentstrings? Most but not all include a space after (resp. before) the comment marker(s). Is there interest in homogenizing (some of) them?

Comment (jamessan) [reply runtime/ftplugin/roc.vim]: I can see value in `commentstring=#\ %s`. I've also seen mention of `comments=b:#`, which I think should only be used if the syntax _requires_ whitespace after the `#`.

Comment (nat-418) [reply runtime/ftplugin/roc.vim]: > I don't know the language's formatting norms but it's probably more common to specify `commenstring` as `#\ %s` for a line comment.
> 
> This is what the Elm ftplugin does, for example.
> 
> https://github.com/ElmCast/elm-vim/blob/4b71facd77297cb33bbb3b14894676cff0a9bd1d/ftplugin/elm.vim#L41

Roc and Elm differ: in Roc, a comment can be written as `# comment` or `## doc comment`, the latter form having some special meaning to the Roc CLI tool. Also, Roc has no multiline strings, as Elm does. What is the best way to represent this if not `#%s`?  I have changed to `#\ %s`, assuming that is the correct answer.

File: src/testdir/test_filetype.vim [hunk 3]
```diff
@@ -595,6 +595,7 @@ def s:GetFilenameChecks(): dict<list<string>>
     rpgle: ['file.rpgle', 'file.rpgleinc'],
     robot: ['file.robot', 'file.resource'],
     robots: ['robots.txt'],
+    roc: ['file.roc'],
```
Comment (zeertzjq) [inline src/testdir/test_filetype.vim:598]: Why did you add a test for filetype detection without implementing it?

Comment (nat-418) [reply src/testdir/test_filetype.vim:598]: Mistake on my part

File: runtime/ftplugin/roc.vim (hunk 2 above)
Comment (dkearns) [reply runtime/ftplugin/roc.vim]: > Sorry for hijacking this discussion, but it struck me as a good (and relevant) opportunity: Is there an official Vim policy on commentstrings? Most but not all include a space after (resp. before) the comment marker(s). Is there interest in homogenizing (some of) them?

I think it would probably be a good idea.

I don't like the lack of consistency throughout the runtime files, in general, but in the past these decisions were intentionally left to the maintainers.

File: runtime/ftplugin/roc.vim [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+" Language: Roc
+" Maintainer: nat-418 <93013864+nat-418@users.noreply.github.com>
+" Latest Revision: 2024-04-5
+
+if exists("b:did_ftplugin")
+  finish
+endif
+let b:did_ftplugin = 1
+
+setlocal comments=:#
```
Comment (dkearns) [inline runtime/ftplugin/roc.vim]: You probably want to use the following so that doc comments are distinctly matched.

`setlocal comments=:##,:#`

Then the appropriate comment character(s) will be used when formatting comments etc.

File: runtime/ftplugin/roc.vim [hunk 5]
```diff
@@ -0,0 +1,14 @@
+" Roc filetype plugin file
+" Language: Roc
+" Maintainer: nat-418 <93013864+nat-418@users.noreply.github.com>
```
Comment (dkearns) [inline runtime/ftplugin/roc.vim:3]: Using a 'real' name would be preferred but I'm aware there's plenty of good reasons not to do so.

@chrisbra, do you care about this?  It was somewhat frowned upon in the past.

Comment (chrisbra) [reply runtime/ftplugin/roc.vim:3]: Yeah, I suppose the email is invalid anyhow? I prefer real-names as well, but won't insist on it.

Comment (nat-418) [reply runtime/ftplugin/roc.vim:3]: I use the github provided email to prevent spam. I use the moniker to avoid weirdos.

Comment (chrisbra) [reply runtime/ftplugin/roc.vim:3]: okay thanks
